outbox.js

```diff
@@ -0,0 +1,219 @@
+class DistributedOutbox extends HTMLElement {
+  constructor() {
+    super();
+    this.renderedItems = new Map(); // Tracks rendered items by ID
+    this.numPosts = 32; // Default value
+    this.page = 1; // Default value
+    this.totalPages = 0; // Keep track of total pages
+  }
+
+  static get observedAttributes() {
+    return ["url", "num-posts", "page"];
+  }
+
+  connectedCallback() {
+    // Use attributes or default values
+    this.numPosts = parseInt(this.getAttribute("num-posts"), 10) || this.numPosts;
+    this.page = parseInt(this.getAttribute("page"), 10) || this.page;
+    this.loadOutbox(this.getAttribute("url"));
+    this.paginationControls();
+  }
+
+  async loadOutbox(outboxUrl) {
+    this.clearContent();
+    for await (const item of this.fetchOutboxItems(outboxUrl)) {
+      const itemKey = item.object;
+      if (itemKey === undefined) {
+        console.error("Item key (object property) is undefined, item:", item);
+        continue; // Skip this item
+      }
+      if (!this.renderedItems.has(itemKey)) {
+        this.renderItem(item);
+        // Mark as rendered by adding to the Map
+        this.renderedItems.set(itemKey, item);
+        // console.log(`Rendered item with key: ${itemKey}`);
+      }
+      // else {
+      //   console.log(`Duplicate item with key: ${itemKey} skipped`);
+      // }
+    }
+  }
+
+  async *fetchOutboxItems(outboxUrl) {
+    if (!outboxUrl) {
+      console.error("No outbox URL provided");
+      return;
+    }
+
+    try {
+      const response = await fetch(outboxUrl);
+      if (!response.ok) {
+        throw new Error(`HTTP error! Status: ${response.status}`);
+      }
+      const outboxData = await response.json();
+
+      this.totalPages = Math.ceil(
+        outboxData.orderedItems.length / this.numPosts
+      );
+      // Prevent page number from going beyond total pages
+      this.page = Math.min(this.page, this.totalPages);
+
+      // Simulate pagination by slicing the items array
+      const startIndex = (this.page - 1) * this.numPosts;
+      const endIndex = startIndex + this.numPosts;
+      const paginatedItems = outboxData.orderedItems.slice(
+        startIndex,
+        endIndex
+      );
+
+      for (const item of paginatedItems) {
+        yield item;
+      }
+    } catch (error) {
+      console.error("Error fetching outbox:", error);
+    }
+  }
+
+  renderItem(item) {
+    const activityElement = document.createElement("distributed-activity");
+    activityElement.type = item.type;
+    activityElement.data = item;
+    this.appendChild(activityElement);
+  }
+
+  paginationControls() {
+    // Attach event listeners for pagination
+    const prevButton = document.getElementById("prevPage");
```

Comment:
These listeners should be attached inside outbox.html since the buttons aren't part of the component.